### ISSUE TITLE
ascii.read inconsistent treatment of data_start and header_start

### ISSUE THREAD

**Issue by bhilbert4** [astropy/astropy#3816]
I ran across some strange behavior while trying to use ascii.read to read in a text file. It appears that lines containing only spaces in the file to be read in are counted differently with respect to the data_start and header_start keyword values. I'm using astropy version 1.0.1.

Here is the file I was trying to read in (myfile.txt):
Data produced from myscript.py
Box numbers and gain values for quadrant 3 of detector NRCA1, calculated from files:
File1.fits
File2.fits
  
The locations and shapes of the boxes in this file are given in: boxmapfile.fits

boxnumber   gain
30002.0   2.04351925985
30000.0   2.05034584049
30003.0   2.05689883435
30001.0   2.03117647546

Note the blank lines after 'File2.fits' and after 'The locations...'
When these are truly blank lines, ascii.read behaves as expected:
    `from astropy.io import ascii`
    `indtable = ascii.read('myfile.txt',header_start=7,data_start=8)`
    `indtable.colnames`
['boxnumber', 'gain']
    `indtable.columns['boxnumber'].data`
array([ 30002.,  30000.,  30003.,  30001.])

But if those two blank lines each contain a space, then ascii.read finds the column headers correctly, but skips the first two lines of data and reads only the last two.
   `from astropy.io import ascii`
   `indtable = ascii.read('myfile.txt',header_start=7,data_start=8)`
`indtable.colnames`
['boxnumber', 'gain']
`indtable.columns['boxnumber'].data`
array([ 30003.,  30001.])

Decreasing the data_start keyword value by 2, I can make ascii.read work correctly in the case where the blank lines contain spaces:
indtable = ascii.read('myfile.txt',header_start=7,data_start=6)
`indtable.colnames`
['boxnumber', 'gain']
`indtable.columns['boxnumber'].data`
array([ 30002.,  30000.,  30003.,  30001.])

This suggests that lines containing only spaces are counted as lines when moving through the file to reach the header, but these same lines are not counted when moving through the file to reach the data.

**Comment by pllim** [comment]
The problem above can be reproduced as follow. Let's consider:
```
>>> lines1 = ['#1', '#C2', ' ', 'data1', 'data2']  # with one blank space on Line 3
>>> lines2 = ['#1', '#C2', '', 'data1', 'data2']  # with empty string on Line 3 
```

`BaseData` treats them as the same, but not `BaseHeader`:
```
>>> from astropy.io.ascii import BaseHeader, BaseData
>>> dat = BaseData()
>>> dat.process_lines(lines1)
['#1', '#C2', 'data1', 'data2']
>>> dat.process_lines(lines2)
['#1', '#C2', 'data1', 'data2']
>>> hdr = BaseHeader()
>>> list(hdr.process_lines(lines1))
['#1', '#C2', ' ', 'data1', 'data2']
>>> list(hdr.process_lines(lines2))
['#1', '#C2', 'data1', 'data2']
```

Therefore, I think `BaseHeader.process_lines()` needs to be fixed. There is no reason to treat `" "` and `""` differently. Also, I don't know why `BaseData.process_lines()` returns a list, but `BaseHeader.process_lines()` returns an iterator; another inconsistency, but not related to this issue.

**Comment by taldcroft** [comment]
This is a repeat of #2654.  There was some concern about breaking existing code that has been set up to work around this "feature".  But it is probably just time to bite the bullet and make the fix.

**Comment by pllim** [comment]
Well, maybe we can add a keyword, something to the effect of "do_not_count_blank_lines", and make that the new default? That way, people who rely on this "feature" can turn it off. In addition, we also add a deprecation warning that says something like "in the next release, we are going to not count blank lines regardless". Just an idea.

**Comment by taldcroft** [comment]
That would work, but maybe if people are already having to update their code to add the `do_not_count_blank_lines=False` then it would be just as easy to fix their `header_start`.

**Comment by pllim** [comment]
Good point, @taldcroft ! :smile: 

Regardless, I am in favor of fixing it. But we probably want some obvious statement somewhere to warn those users.

**Comment by hamogu** [comment]
I want to reiterate what I said in https://github.com/astropy/astropy/issues/2654#issuecomment-60504246:

I suggest:
Positive integers for `data_start` and `header_start` do whatever we decide on here, but negative integers just count the true number of lines (even empty ones and lines with spaces). [Or the other way around.]

I recently wanted to read a table in a very big log file (*). I open it in emacs and I know that the tables starts  in line 35298, but it's impossible to use "header_start" or "data_start" because > 2000 or those lines are empty and I don't really want to go through and count them. I ended up with copy & paste of the relevant part into a new file, but that's exactly what I would have liked to avoid.

(*) It loggs some simulation output, including tables for the atmospheric structure at every step. And it's legacy code and not even my code - I cannot change the output file format, I need to read what I have.

**Comment by taldcroft** [comment]
I'm not so excited about the negative value convention:

- Negative values are currently legal and indicate the index from the end like normal lists / arrays.
- This convention wouldn't be obvious to someone reading the code.

I would rather add a keyword like `absolute_start_end=False/True`.  (Not the perfect name, but we could find something).

**Comment by hamogu** [comment]
I did not even know that negative numbers are currently legal and I've thought *a lot* about the header and data start lines when I wrote some of the previous PRs relayed to this issue. I wonder how many users know and use this feature...

**Comment by taldcroft** [comment]
Closed by #2654.

@hamogu - if you want to pursue some way of having the counts be independent of whitespace / comments then feel free to open an issue.  From the code perspective I didn't see a trivial way to do this.
